File: Application/src/main/java/com/shareyourproxy/api/rx/command/AddGroupChannelAndPublicCommand.java
package com.shareyourproxy.api.rx.command;

import android.app.Service;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.shareyourproxy.api.domain.model.Channel;
import com.shareyourproxy.api.domain.model.GroupToggle;
import com.shareyourproxy.api.domain.model.User;
import com.shareyourproxy.api.rx.RxGroupChannelSync;
import com.shareyourproxy.api.rx.RxUserChannelSync;
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback;
import com.shareyourproxy.api.rx.command.eventcallback.UserGroupAddedEventCallback;

import java.util.ArrayList;

import static com.shareyourproxy.api.domain.factory.ChannelFactory.createPublicChannel;

/**
 * Update users groups channel.
 */
public class AddGroupChannelAndPublicCommand extends BaseCommand {
    public static final Parcelable.Creator<AddGroupChannelAndPublicCommand> CREATOR =
        new Parcelable.Creator<AddGroupChannelAndPublicCommand>() {
            @Override
            public AddGroupChannelAndPublicCommand createFromParcel(Parcel in) {
                return new AddGroupChannelAndPublicCommand(in);
            }

            @Override
            public AddGroupChannelAndPublicCommand[] newArray(int size) {
                return new AddGroupChannelAndPublicCommand[size];
            }
        };
    private final static java.lang.ClassLoader CL =
        AddGroupChannelAndPublicCommand.class.getClassLoader();
    public final User user;
    public final Channel channel;
    public final ArrayList<GroupToggle> groups;

    /**
     * Public constructor.
     *
     * @param user    Logged in user
     * @param groups  selected groups to add channel to
     * @param channel this events group
     */
    public AddGroupChannelAndPublicCommand(
        @NonNull User user, @NonNull ArrayList<GroupToggle> groups, @NonNull Channel channel) {
        this.user = user;
        this.groups = groups;
        this.channel = channel;
    }

    private AddGroupChannelAndPublicCommand(Parcel in) {
        this((User) in.readValue(CL), (ArrayList<GroupToggle>) in.readValue(CL),
            (Channel) in.readValue(CL));
    }

    @Override
    public EventCallback execute(Service service) {
        Channel publicChannel = createPublicChannel(channel, true);
        UserGroupAddedEventCallback updatedUser =
            RxGroupChannelSync.INSTANCE.addUserGroupsChannel(service, user, groups, publicChannel);
        return RxUserChannelSync.INSTANCE.saveUserChannel(service, updatedUser.user, channel, publicChannel);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(user);
        dest.writeValue(groups);
        dest.writeValue(channel);
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/api/rx/command/AddGroupChannelAndPublicCommand.kt
package com.shareyourproxy.api.rx.command

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import com.shareyourproxy.api.domain.factory.ChannelFactory.createPublicChannel
import com.shareyourproxy.api.domain.model.Channel
import com.shareyourproxy.api.domain.model.GroupToggle
import com.shareyourproxy.api.domain.model.User
import com.shareyourproxy.api.rx.RxGroupChannelSync
import com.shareyourproxy.api.rx.RxUserChannelSync.saveUserChannel
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback
import java.util.*

/**
 * Update users groups channel.
 * @param user    Logged in user
 * @param groups  selected groups to add channel to
 * @param channel this events group
 */
internal final class AddGroupChannelAndPublicCommand(val user: User, val groups: ArrayList<GroupToggle>, val channel: Channel) : BaseCommand() {
    @Suppress("UNCHECKED_CAST")
    private constructor(parcel: Parcel) : this(parcel.readValue(CL) as User, parcel.readValue(CL) as ArrayList<GroupToggle>, parcel.readValue(CL) as Channel)

    override fun execute(context: Context): EventCallback {
        val publicChannel = createPublicChannel(channel, true)
        val updatedUser = RxGroupChannelSync.addUserGroupsChannel(context, user, groups, publicChannel)
        return saveUserChannel(context, updatedUser.user, channel, publicChannel)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(user)
        dest.writeValue(groups)
        dest.writeValue(channel)
    }

    companion object {
        private val CL = AddGroupChannelAndPublicCommand::class.java.classLoader
        @JvmField val CREATOR: Parcelable.Creator<AddGroupChannelAndPublicCommand> = object : Parcelable.Creator<AddGroupChannelAndPublicCommand> {
            override fun createFromParcel(parcel: Parcel): AddGroupChannelAndPublicCommand {
                return AddGroupChannelAndPublicCommand(parcel)
            }

            override fun newArray(size: Int): Array<AddGroupChannelAndPublicCommand?> {
                return arrayOfNulls(size)
            }
        }
    }
}
